src/integration-test/java/com/commercetools/sync/integration/sdk2/ctpprojectsource/products/ProductReferenceResolverIT.java

```diff
@@ -0,0 +1,432 @@
+import com.commercetools.api.models.product.ProductVariantDraftBuilder;
+import com.commercetools.api.models.product_type.ProductType;
+import com.commercetools.api.models.state.State;
+import com.commercetools.api.models.state.StateReference;
+import com.commercetools.api.models.state.StateReferenceBuilder;
+import com.commercetools.api.models.state.StateTypeEnum;
+import com.commercetools.api.models.tax_category.TaxCategory;
+import com.commercetools.sync.sdk2.commons.exceptions.ReferenceResolutionException;
+import com.commercetools.sync.sdk2.commons.utils.CaffeineReferenceIdToKeyCacheImpl;
+import com.commercetools.sync.sdk2.commons.utils.ReferenceIdToKeyCache;
+import com.commercetools.sync.sdk2.products.ProductSync;
+import com.commercetools.sync.sdk2.products.ProductSyncOptions;
+import com.commercetools.sync.sdk2.products.ProductSyncOptionsBuilder;
+import com.commercetools.sync.sdk2.products.helpers.ProductSyncStatistics;
+import com.commercetools.sync.sdk2.products.utils.ProductTransformUtils;
+import io.vrap.rmf.base.client.ApiHttpResponse;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.concurrent.CompletionException;
+import javax.annotation.Nonnull;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class ProductReferenceResolverIT {
+  private static ProductType productTypeSource;
+  private static ProductType noKeyProductTypeSource;
+  private static ProductType productTypeSourceWithReferenceTypeVariantAttribute;
+
+  private static TaxCategory oldTaxCategory;
+  private static State oldProductState;
+  private static Customer oldCustomer;
+  private static List<CategoryReference> categoryReferencesWithIds;
+  private ProductSync productSync;
+  private List<String> errorCallBackMessages;
+  private List<String> warningCallBackMessages;
+  private List<Throwable> errorCallBackExceptions;
+  private ReferenceIdToKeyCache referenceIdToKeyCache;
+
+  /**
+   * Delete all product related test data from target and source projects. Then creates custom types
+   * for both CTP projects categories.
+   */
+  @BeforeAll
+  static void setup() {
+    ensureCategoriesCustomType(
+        OLD_CATEGORY_CUSTOM_TYPE_KEY,
+        Locale.ENGLISH,
+        OLD_CATEGORY_CUSTOM_TYPE_NAME,
+        CTP_TARGET_CLIENT);
+    ensureCategoriesCustomType(
+        OLD_CATEGORY_CUSTOM_TYPE_KEY,
+        Locale.ENGLISH,
+        OLD_CATEGORY_CUSTOM_TYPE_NAME,
+        CTP_SOURCE_CLIENT);
+
+    ensureCategories(CTP_TARGET_CLIENT, getCategoryDrafts(null, 2));
+    categoryReferencesWithIds =
+        getReferencesWithIds(ensureCategories(CTP_SOURCE_CLIENT, getCategoryDrafts(null, 2)));
+
+    ensureProductType(PRODUCT_TYPE_RESOURCE_PATH, CTP_TARGET_CLIENT);
+    //    ensureProductType(PRODUCT_TYPE_NO_KEY_RESOURCE_PATH, CTP_TARGET_CLIENT);
```

Comment:
Remove this comment

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/937/commits/ce838b0eb9845530ca81fce9fc944139cebf80d0